File: app/src/main/java/fr/corenting/edcompanion/views/GraphMarkerView.java
package fr.corenting.edcompanion.views;

import android.content.Context;
import android.widget.TextView;

import com.github.mikephil.charting.components.MarkerView;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.utils.MPPointF;

import org.threeten.bp.LocalDateTime;
import org.threeten.bp.ZoneOffset;
import org.threeten.bp.format.DateTimeFormatter;
import org.threeten.bp.format.FormatStyle;

import fr.corenting.edcompanion.R;
import fr.corenting.edcompanion.models.FactionChartEntryData;

public class GraphMarkerView extends MarkerView {

    private final DateTimeFormatter dateFormatter;

    private TextView factionNameTextView;
    private TextView influenceTextView;
    private TextView stateTextView;
    private TextView updateDateTextView;

    public GraphMarkerView(Context context) {
        super(context, R.layout.view_marker);

        factionNameTextView = findViewById(R.id.factionNameTextView);
        influenceTextView = findViewById(R.id.influenceTextView);
        stateTextView = findViewById(R.id.stateTextView);
        updateDateTextView = findViewById(R.id.updateDateTextView);

        dateFormatter = DateTimeFormatter.ofLocalizedDate(FormatStyle.SHORT);
    }

    @Override
    public void refreshContent(Entry e, Highlight highlight) {
        FactionChartEntryData historyItem = (FactionChartEntryData) e.getData();
        factionNameTextView.setText(historyItem.getName());

        influenceTextView.setText(String.format("%s %%", (int) (historyItem.getInfluence() * 100)));
        stateTextView.setText(historyItem.getState());

        LocalDateTime date = LocalDateTime.ofInstant(historyItem.getUpdateDate(), ZoneOffset.UTC);
        updateDateTextView.setText(dateFormatter.format(date));

        super.refreshContent(e, highlight);
    }

    private MPPointF mOffset;

    @Override
    public MPPointF getOffset() {

        if (mOffset == null) {
            // center the marker horizontally and vertically
            mOffset = new MPPointF(-(getWidth() / 2), -getHeight());
        }

        return mOffset;
    }
}

File: app/src/main/java/fr/corenting/edcompanion/views/GraphMarkerView.kt
package fr.corenting.edcompanion.views

import android.content.Context
import android.graphics.Canvas
import android.view.LayoutInflater
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.components.MarkerView
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.highlight.Highlight
import fr.corenting.edcompanion.R
import fr.corenting.edcompanion.databinding.ViewMarkerBinding
import fr.corenting.edcompanion.models.FactionChartEntryData
import fr.corenting.edcompanion.utils.DateUtils
import org.threeten.bp.LocalDateTime
import org.threeten.bp.ZoneOffset
import org.threeten.bp.format.DateTimeFormatter
import org.threeten.bp.format.FormatStyle
import java.util.Locale


class GraphMarkerView(context: Context?) : MarkerView(context, R.layout.view_marker) {

    private var binding: ViewMarkerBinding =
        ViewMarkerBinding.inflate(LayoutInflater.from(context), this, true)

    private val dateFormatter: DateTimeFormatter =
        DateTimeFormatter.ofLocalizedDate(FormatStyle.SHORT)
    private val userLocale: Locale = DateUtils.getCurrentLocale(context)

    override fun refreshContent(e: Entry, highlight: Highlight) {
        val (name, state, updateDate, influence) = e.data as FactionChartEntryData
        binding.factionNameTextView.text = name
        binding.influenceTextView.text = String.format(
            userLocale,
            "%.2f %%", influence * 100
        )
        binding.stateTextView.text = state
        val date = LocalDateTime.ofInstant(updateDate, ZoneOffset.UTC)
        binding.updateDateTextView.text = dateFormatter.format(date)
        super.refreshContent(e, highlight)
    }

    fun setChartView(chart: LineChart) {
        chartView = chart
    }

    override fun draw(canvas: Canvas, posX: Float, posY: Float) {
        super.draw(canvas, posX, posY)
        getOffsetForDrawingAtPoint(posX, posY)
    }
}